cpp/include/Ice/Exception.h

```diff
@@ -1,112 +1,49 @@
-//
-// Copyright (c) ZeroC, Inc. All rights reserved.
-//
+// Copyright (c) ZeroC, Inc.
 
 #ifndef ICE_EXCEPTION_H
 #define ICE_EXCEPTION_H
 
 #include "Config.h"
-#include "ValueF.h"
 
 #include <exception>
-#include <memory>
 #include <ostream>
-#include <string>
-#include <vector>
 
 namespace Ice
 {
-    /**
-     * Abstract base class for all Ice exceptions. It has only two derived classes: LocalException and UserException.
-     * \headerfile Ice/Ice.h
-     */
+    /// Abstract base class for all Ice exceptions. It has only two derived classes: LocalException and UserException.
+    /// \headerfile Ice/Ice.h
     class ICE_API Exception : public std::exception
     {
     public:
-        /**
-         * Constructs an exception with a default message.
-         * @param file The file where this exception is constructed. This C string is not copied.
-         * @param line The line where this exception is constructed.
-         */
-        Exception(const char* file, int line) noexcept;
-
-        /**
-         * Constructs an exception.
-         * @param file The file where this exception is constructed. This C string is not copied.
-         * @param line The line where this exception is constructed.
-         * @param message The error message adopted by this exception and returned by what().
-         */
-        Exception(const char* file, int line, std::string message);
-
-        /**
-         * Copy constructor.
-         * @param other The exception to copy.
-         */
-        Exception(const Exception& other) noexcept = default;
-
-        /**
-         * Assignment operator.
-         * @param rhs The exception to assign.
-         */
-        Exception& operator=(const Exception& rhs) noexcept = default;
-
-        /**
-         * Returns the error message of this exception.
-         * @return The error message.
-         */
-        [[nodiscard]] const char* what() const noexcept override;
-
-        /**
-         * Returns the type ID of this exception. This corresponds to the Slice type ID for Slice-defined exceptions,
-         * and to a similar fully scoped name for other exceptions. For example "::Ice::CommunicatorDestroyedException".
-         * @return The type ID of this exception
-         */
+        Exception() = default;
+        Exception(const Exception&) = default;
+        Exception& operator=(const Exception&) = default;
+
+        // Need out-of-line virtual function to avoid weak vtable, which in turn requires the default constructor,
```

Comment:
Weak vtables break `dynamic_cast` and `catch(const XxxException&)`, and it's typically hard to find since we see these errors only at runtime.  This particular weak vtable breaks the Ice/background test when it catches `const Ice::Exception&` on macos.